### DISCUSSION TITLE
FFmpeg SCTE-35 developments

### DISCUSSION THREAD

**Discussion by bbgdzxng1** [futzu/threefive#86] [General]
Hi Adrian,

Feel free to close this as you see fit, but I just wanted to flag that there is an interesting patch in FFmpeg devel mailing list regarding SCTE-35 data tracks that has recently appeared.  https://patchwork.ffmpeg.org/project/ffmpeg/list/?series=9118 and http://ffmpeg.org/pipermail/ffmpeg-devel/2023-June/310798.html.

If the patch progresses (and there is no guarantee where FFmpeg is concerned, especially involving anything regarding data tracks), your userbase may start trying to use threefive against streams that have been processed by this patchset, so I wanted give you the heads-up.  As the SCTE authority in the open source community, your opinion is likely to be valuable in case the FFmpeg patchset goes off-piste.  If there is stuff that you think FFmpeg could do better prior to your suite having to accommodate the nuances of FFmpeg, you may be able to steer the people.  The developer's contact details are in the ticket.

It looks like https://patchwork.ffmpeg.org/project/ffmpeg/patch/1686953578-18843-4-git-send-email-dheitmueller@ltnglobal.com/ will correctly set the FourCC as `CUEI`, which is a good thing.  I'm no coder, but it looks like since `AV_CODEC_ID_SCTE_35` is a known, emumerated string, it should preserve the data type as 0x86.  Of course, six2scte35 is a great solution, but one which exists to accommodate the nuances of another tool.

There's a few other tweaks in there - bitstream filters - which could be interesting.

Hopefully threefive will be the tool that is used to validate the patchset.  It is certainly the best out there.  I hope they also test with TS Duck.

Like many SCTE-35 patches in FFmpeg, this may be one that languishes untouched, but hopefully this time there will be some momentum.

Hope you are well, sir.  No reply necessary - feel free to close (and ignore).

**Comment by futzu** [comment]
I know the cat who did the patch, 
that's who I used to ask my SCTE-35 questions. 

I don't do well when I try to contribute to other projects, I've had at least a dozen patches rejected in the last year.

 I am officially banned from python after I submitted PEP patches and expressed my opinion on a few things, 
they even blocked me from their repos over a 3 line patch. That's just sad. 

I started  M3ufu after the M3u8 guys  rejected my patches for SCTE-35.
Speaking of m3ufu, I did not mean to upset you, I guess I just wasn't getting what you were telling me, 
but I really value what you have to say man, If you look at m3ufu, you'll see a lot of updates inspired by you. 
I miss you being around man.

**Comment by bbgdzxng1** [comment]
All good.  Lots of respect from here-to-you for all your work.  I check your github commits weekly, without fail.  I appreciate that github is not a social network, but I genuinely welled-up when I read what you were going through.  Stay strong, brother.  I wanted to reach out based on your recent status update.  So sorry, man.

Your PEP comments in the README were (seriously) funny.  I've also watched your interactions trying to help a cable-company's commits in their own repo, requiring legal docs.  In neither of these interactions, you were not in the wrong.  Don't be disheartened - you are the authority on SCTE-35, and your toolset is the best out there.

I don't know the developer who did the FFmpeg thing, but I'm very pleased you do.  Great to hear he or she knows of your work and you are aware of theirs.  That's great that you know the dudette/dude.

I never left - just didn't want to p*ss you off.  All cool my end. ❤️

**Comment by futzu** [comment]
I haven't talk to Dimitri, the guy who did the patch,  since I quit working for LTN. 

 I am really surprised that nobody else has done a work around for it ,. That was one of the first things I did  to threefive was setup to handle ffmpeg "bin data". A lot of SCTE-35 parsers require the registration descriptors to be present as well, threefive does not. 

I just don't believe in throwing exceptions unless I absolutely have to do it. It's called non-validating parsing, I got it from expat. Nobody wants to see an error code.
Recently, I started seeing improperly padded base64 strings from Amazon,  Base64 strings are  supposed to be a multiple of 4 in length, so I added a padding checker.

**Comment by futzu** [comment]
By the way, I've been working on cuei lately, a company called [tunein](https://tunein.com) contracted me to add encoding to cuei. These guys are super cool. 
Most companies make me sign a contract stating that none of the code I write for them will ever show up in my open source projects, tunein, insisted that all the code I write for them is open source. 
I thought it was odd at first, but once I thought about it, it does make a lot of sense. Once I've finished with them, I'll keep updating cuei, and that keeps them up to date as well.  

I am probably about to retire from writing code, I am an economist, I don't have any computer training.

**Comment by bbgdzxng1** [comment]
That is very cool of Tunein.

My two favorites in the suite remain threefive and m3ufu.  I gotta say that m3ufu has become a really useful tool - not just for CUEs and SCTE, but as a general m3u8 parser.  You've got a great tool there.

I do hope that if you do return to the world of John Maynard Keynes that your projects are recognized for their contribution to Free Ad Supported TV.

[ As I get older, I'm starting to believe that Keynes was onto something with his "dig a hole, fill a hole" - he may not have been as crazy as it sounded.  We're in a testing time where some of the ground rules of modern economic theory is being tested. ]

Stay cool, my friend.
